File: engine/bg-core/src/record.rs
//! Match records: the seed plus a log of every turn, and replay.
//!
//! Dice are never stored as the source of truth: [`replay`] re-derives every
//! roll from `Record::seed` with [`DiceRng`] and checks that the logged
//! dice, players and plays agree with what the rules produce. A record whose
//! log disagrees with the seed or the rules is rejected, so a stored record
//! is a proof of the game that was played.
//!
//! JSON: `{ "seed": 123456789, "length": 7, "rules": Rules, "turns": [Turn] }`
//! with `Turn` = `{ "player": "white", "dice": Dice|null, "action":
//! "roll"|"move"|"double"|"take"|"drop"|"resign", "play": "24/18 13/10"|null,
//! "resignPoints": u8|null }`; `play` is in notation relative to `player`.
//!
//! `seed` is a bare JSON integer and records travel through JavaScript, whose
//! `JSON.parse` rounds integers above 2^53 − 1. A seed must therefore not
//! exceed [`MAX_SEED`]; [`replay`] rejects a larger one up front rather than
//! failing later on dice derived from a drifted seed.

use serde::{Deserialize, Serialize};

use crate::game::{Phase, ResultKind, Rules};
use crate::match_play::MatchState;
use crate::{Dice, DiceRng, Play, Player, RulesError, parse_play};

/// What a logged turn did. JSON: lowercase.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Action {
    /// The player rolled (`dice` set); the opening roll is logged by its winner.
    Roll,
    /// The player played `play` (`dice` set to the roll played).
    Move,
    /// The player offered a double.
    Double,
    /// The player accepted a double.
    Take,
    /// The player declined a double.
    Drop,
    /// The player resigned, conceding `resign_points` points.
    Resign,
}

/// One logged action.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Turn {
    /// Who acted.
    pub player: Player,
    /// The dice rolled (`Roll`) or played (`Move`); `None` otherwise.
    pub dice: Option<Dice>,
    /// What happened.
    pub action: Action,
    /// The play in standard notation relative to `player` (`Move` only;
    /// `""` for a forfeited turn).
    pub play: Option<String>,
    /// Points conceded (`Resign` only).
    pub resign_points: Option<u8>,
}

impl Turn {
    fn bare(player: Player, action: Action) -> Self {
        Self {
            player,
            dice: None,
            action,
            play: None,
            resign_points: None,
        }
    }

    /// `player` rolled `dice` (for the opening roll, `player` is its winner).
    #[must_use]
    pub fn roll(player: Player, dice: Dice) -> Self {
        Self {
            dice: Some(dice),
            ..Self::bare(player, Action::Roll)
        }
    }

    /// `player` played `play` with `dice`.
    #[must_use]
    pub fn mv(player: Player, dice: Dice, play: &Play) -> Self {
        Self {
            dice: Some(dice),
            play: Some(play.to_string()),
            ..Self::bare(player, Action::Move)
        }
    }

    /// `player` offered a double.
    #[must_use]
    pub fn double(player: Player) -> Self {
        Self::bare(player, Action::Double)
    }

    /// `player` accepted a double.
    #[must_use]
    pub fn take(player: Player) -> Self {
        Self::bare(player, Action::Take)
    }

    /// `player` declined a double.
    #[must_use]
    pub fn drop(player: Player) -> Self {
        Self::bare(player, Action::Drop)
    }

    /// `player` resigned, conceding `points`.
    #[must_use]
    pub fn resign(player: Player, points: u8) -> Self {
        Self {
            resign_points: Some(points),
            ..Self::bare(player, Action::Resign)
        }
    }
}

/// The largest seed a [`Record`] may carry: JavaScript's
/// `Number.MAX_SAFE_INTEGER` (2^53 − 1), so that `seed` survives
/// `JSON.parse` in every binding unchanged.
pub const MAX_SEED: u64 = (1 << 53) - 1;

/// A complete (or partial) match record.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Record {
    /// Seed of the [`DiceRng`] that produced every roll; at most
    /// [`MAX_SEED`] (see the module documentation).
    pub seed: u64,
    /// Match length (`0` = money / single game).
    pub length: u8,
    /// Rules in force for every game.
    pub rules: Rules,
    /// The turns, in order.
    pub turns: Vec<Turn>,
}

impl Record {
    /// An empty record for a match with these parameters.
    #[must_use]
    pub fn new(seed: u64, length: u8, rules: Rules) -> Self {
        Self {
            seed,
            length,
            rules,
            turns: Vec::new(),
        }
    }
}

/// Replays `record` from its seed and returns the resulting [`MatchState`].
///
/// Every roll is re-derived from the seed; every logged `dice`, `player` and
/// `play` must agree with the rules. Whenever an action leaves the game
/// finished, [`MatchState::finish_game`] is called at once — a live driver
/// must do the same for its final state to compare equal.
///
/// # Errors
///
/// [`RulesError::Parse`] when `seed` exceeds [`MAX_SEED`], when the log
/// disagrees with the seed or names the wrong player, when a `Move` has no
/// play or a `Resign` no points, or when a play does not parse; the game's
/// own errors ([`RulesError::IllegalPlay`], [`RulesError::WrongPhase`],
/// [`RulesError::NotAllowed`]) when an action is not legal where it appears.
pub fn replay(record: &Record) -> Result<MatchState, RulesError> {
    if record.seed > MAX_SEED {
        return Err(mismatch(format!(
            "seed {} exceeds the maximum {MAX_SEED} (2^53 - 1, JavaScript's safe integer range)",
            record.seed
        )));
    }
    let mut rng = DiceRng::from_seed(record.seed);
    let mut m = MatchState::new(record.length, record.rules);
    for (i, turn) in record.turns.iter().enumerate() {
        replay_turn(&mut m, &mut rng, turn).map_err(|e| annotate(i, e))?;
        if m.game.phase == Phase::Finished {
            m.finish_game();
        }
    }
    Ok(m)
}

/// Prefixes `Parse` errors with the turn index; other errors pass through.
fn annotate(i: usize, e: RulesError) -> RulesError {
    match e {
        RulesError::Parse(msg) => RulesError::Parse(format!("turn {i}: {msg}")),
        other => other,
    }
}

fn mismatch(msg: String) -> RulesError {
    RulesError::Parse(msg)
}

fn expect_player(turn: &Turn, expected: Option<Player>) -> Result<(), RulesError> {
    if expected == Some(turn.player) {
        Ok(())
    } else {
        Err(mismatch(format!(
            "logged player {:?} but {:?} is to act",
            turn.player, expected
        )))
    }
}

fn expect_dice(turn: &Turn, actual: Dice) -> Result<(), RulesError> {
    match turn.dice {
        Some(logged) if logged == actual => Ok(()),
        Some(logged) => Err(mismatch(format!(
            "logged dice {}-{} but the seed gives {}-{}",
            logged.hi, logged.lo, actual.hi, actual.lo
        ))),
        None => Err(mismatch("turn has no dice".into())),
    }
}

fn replay_turn(m: &mut MatchState, rng: &mut DiceRng, turn: &Turn) -> Result<(), RulesError> {
    match turn.action {
        Action::Roll if m.game.phase == Phase::OpeningRoll => {
            let dice = m.game.opening_roll(rng);
            expect_player(turn, m.game.on_roll)?;
            expect_dice(turn, dice)
        }
        Action::Roll => {
            expect_player(turn, m.game.on_roll)?;
            let dice = m.game.roll(rng)?;
            expect_dice(turn, dice)
        }
        Action::Move => {
            expect_player(turn, m.game.on_roll)?;
            if let Some(dice) = m.game.dice {
                expect_dice(turn, dice)?;
            }
            let notation = turn
                .play
                .as_deref()
                .ok_or_else(|| mismatch("move turn has no play".into()))?;
            let play = parse_play(notation)?;
            m.game.play(&play)
        }
        Action::Double => {
            expect_player(turn, m.game.on_roll)?;
            if !m.cube_allowed() {
                return Err(RulesError::NotAllowed(
                    "the cube is out of play in the Crawford game",
                ));
            }
            m.game.double()
        }
        Action::Take => {
            expect_player(turn, m.game.on_roll.map(Player::opponent))?;
            m.game.take()
        }
        Action::Drop => {
            expect_player(turn, m.game.on_roll.map(Player::opponent))?;
            m.game.drop()
        }
        Action::Resign => {
            expect_player(turn, m.game.on_roll)?;
            let points = turn
                .resign_points
                .ok_or_else(|| mismatch("resign turn has no points".into()))?;
            let kind = resign_kind(points, m.game.cube.value)?;
            m.game.resign(kind)?;
            let awarded = m.game.result.map_or(0, |r| r.points);
            if awarded == points {
                Ok(())
            } else {
                Err(mismatch(format!(
                    "logged resignation of {points} points but the rules award {awarded}"
                )))
            }
        }
    }
}

/// The kind conceded when `points` are resigned at `cube` (`points` must be
/// 1, 2 or 3 times the cube).
fn resign_kind(points: u8, cube: u8) -> Result<ResultKind, RulesError> {
    let kind = [
        ResultKind::Single,
        ResultKind::Gammon,
        ResultKind::Backgammon,
    ]
    .into_iter()
    .find(|k| k.multiplier().saturating_mul(cube) == points);
    kind.ok_or_else(|| {
        mismatch(format!(
            "resignation of {points} points is not 1, 2 or 3 times the cube ({cube})"
        ))
    })
}

File: web/src/game/record.ts
/**
 * Append helpers for the match record — the single source of truth of a bot
 * game. Every action becomes exactly one `Turn` shaped as `bg_core::Record`
 * expects (`engine/bg-core/src/record.rs`); the store then re-derives the
 * `MatchState` with the engine's `replay`, which verifies dice, players and
 * plays. Nothing here decides legality.
 */

import type {
  Cube,
  Dice,
  MatchContext,
  MatchState,
  Move,
  Play,
  Player,
  Record as GameRecord,
  ResultKind,
  Rules,
  Turn,
} from "@/engine/types";

import { DiceRng, MAX_SEED, assertSeed, type DieSource } from "./dice";

export { MAX_SEED };

/** `Rules::money()` for a single game (`length` 0), `Rules::match_play()` otherwise. */
export function defaultRules(length: number): Rules {
  return length === 0
    ? { jacoby: true, beavers: false, autoDoubles: false }
    : { jacoby: false, beavers: false, autoDoubles: false };
}

/** An empty record for a match (or, with `length` 0, a single game). */
export function newRecord(seed: number, length: number, rules: Rules = defaultRules(length)): GameRecord {
  assertSeed(seed);
  if (!Number.isInteger(length) || length < 0 || length > 255) {
    throw new RangeError(`invalid match length ${length}: must be an integer in 0..=255`);
  }
  return { seed, length, rules, turns: [] };
}

/** A new record with `turns` appended; the input is left untouched. */
export function appendTurn(record: GameRecord, ...turns: Turn[]): GameRecord {
  return { ...record, turns: [...record.turns, ...turns] };
}

function bare(player: Player, action: Turn["action"]): Turn {
  return { player, dice: null, action, play: null, resignPoints: null };
}

/** `player` rolled `dice`; for the opening roll `player` is its winner. */
export function rollTurn(player: Player, dice: Dice): Turn {
  return { ...bare(player, "roll"), dice };
}

/** `player` played `notation` (the engine's own, relative to `player`; `""` = no legal move) with `dice`. */
export function moveTurn(player: Player, dice: Dice, notation: string): Turn {
  return { ...bare(player, "move"), dice, play: notation };
}

export function doubleTurn(player: Player): Turn {
  return bare(player, "double");
}

export function takeTurn(player: Player): Turn {
  return bare(player, "take");
}

export function dropTurn(player: Player): Turn {
  return bare(player, "drop");
}

/** `player` resigned, conceding `points` (see `resignPoints`). */
export function resignTurn(player: Player, points: number): Turn {
  return { ...bare(player, "resign"), resignPoints: points };
}

/**
 * The opening roll exactly as `GameState::opening_roll` draws it: White's
 * die first, then Black's; a tie is drawn again (auto-doubles only touch the
 * cube, which the engine applies during replay); the higher die's owner is
 * on roll and the turn is logged in their name.
 */
export function openingRollTurn(rng: DieSource): Turn {
  for (;;) {
    const white = rng.rollOne();
    const black = rng.rollOne();
    if (white === black) {
      continue;
    }
    const winner: Player = white > black ? "white" : "black";
    return rollTurn(winner, { hi: Math.max(white, black), lo: Math.min(white, black) });
  }
}

/**
 * Tie re-draws tolerated while matching an opening roll. Ties are re-drawn
 * until the dice differ, so a valid record never needs more than a handful;
 * the bound keeps a corrupt record from being searched for ever.
 */
const MAX_OPENING_TIES = 64;

const sameDice = (a: Dice, b: Dice): boolean => a.hi === b.hi && a.lo === b.lo;

/**
 * The dice stream positioned just after the last roll of `record`, for
 * resuming a stored game: `replay` verifies logged dice but never says where
 * the stream stands, so the record is walked the way the engine consumed it.
 * Every roll turn consumed one pair; an opening roll additionally consumed
 * one pair per tie (`GameState::opening_roll`). A tie is never logged as an
 * opening roll and a regular roll always equals the pair drawn for it, so
 * a drawn pair that is a tie yet differs from the logged dice can only be an
 * opening re-roll. Throws when the logged dice cannot follow from the seed
 * (the engine's `replay` would reject the record too).
 */
export function diceStreamAfter(record: GameRecord): DiceRng {
  const rng = new DiceRng(record.seed);
  record.turns.forEach((turn, index) => {
    if (turn.action !== "roll") {
      return;
    }
    const logged = turn.dice;
    if (!logged) {
      throw new Error(`turn ${index}: roll turn has no dice`);
    }
    for (let ties = 0; ; ties++) {
      const drawn = rng.roll();
      if (sameDice(drawn, logged)) {
        return;
      }
      if (drawn.hi !== drawn.lo || ties >= MAX_OPENING_TIES) {
        throw new Error(
          `turn ${index}: logged dice ${logged.hi}-${logged.lo} but the seed gives ${drawn.hi}-${drawn.lo}; the stored record does not follow its seed`,
        );
      }
    }
  });
  return rng;
}

const KIND_MULTIPLIER: { readonly [K in ResultKind]: number } = { single: 1, gammon: 2, backgammon: 3 };

/** `kind` × `cubeValue`, with no rule applied (see `concededPoints` for what a resignation must log). */
export function resignPoints(kind: ResultKind, cubeValue: number): number {
  return KIND_MULTIPLIER[kind] * cubeValue;
}

/**
 * The points the engine awards for a resignation as `kind` in `game` —
 * `GameState::finish` in `engine/bg-core/src/game.rs`: under the Jacoby rule
 * a gammon or backgammon counts as a single game while the cube is still
 * centred (no double was offered), then the multiplier applies to the cube.
 * `replay` rejects a resign turn whose logged points differ from this.
 */
export function concededPoints(kind: ResultKind, game: { rules: Pick<Rules, "jacoby">; cube: Cube }): number {
  const effective: ResultKind = game.rules.jacoby && game.cube.owner === null ? "single" : kind;
  return resignPoints(effective, game.cube.value);
}

/** The kind whose multiplier × `cubeValue` equals `points`, or `null`. */
export function resultKindFor(points: number, cubeValue: number): ResultKind | null {
  const kinds: ResultKind[] = ["single", "gammon", "backgammon"];
  return kinds.find((k) => KIND_MULTIPLIER[k] * cubeValue === points) ?? null;
}

export function opponent(player: Player): Player {
  return player === "white" ? "black" : "white";
}

/** `bg_bot::MatchContext` from `player`'s side of `match`. */
export function matchContextFor(match: MatchState, player: Player): MatchContext {
  const other = opponent(player);
  const away = (p: Player): number => (match.length === 0 ? 0 : Math.max(0, match.length - match.score[p]));
  const owner = match.game.cube.owner;
  return {
    length: match.length,
    myAway: away(player),
    theirAway: away(other),
    crawford: match.crawford,
    postCrawford: match.postCrawford,
    cube: match.game.cube.value,
    cubeOwnerIsMe: owner === null ? null : owner === player,
  };
}

/** A fresh 53-bit seed from the platform CSPRNG (`crypto.getRandomValues`). */
export function randomSeed(): number {
  const words = new Uint32Array(2);
  globalThis.crypto.getRandomValues(words);
  return (words[0] & 0x1f_ffff) * 2 ** 32 + words[1];
}

/**
 * The seed handed to `choosePlay` for the bot's next decision: derived from
 * the record so a game replays identically for the same `seed`, distinct per
 * turn, and within the engine's safe-integer bound.
 */
export function botSeed(record: GameRecord): number {
  const step = 1_000_003;
  return (record.seed + step * (record.turns.length + 1)) % (MAX_SEED + 1);
}

const sameSquare = (a: Move, b: Move): boolean => a.from === b.from && a.to === b.to;

/**
 * The moves of `play` not yet in `pending`, matched by `from`/`to` as a
 * multiset (order and hit flags are ignored: a person may enter the moves
 * of a play in any order). Returns `null` when `pending` is not a
 * sub-multiset of `play.moves`.
 */
function unmatched(pending: readonly Move[], play: Play): Move[] | null {
  const rest = [...play.moves];
  for (const move of pending) {
    const i = rest.findIndex((m) => sameSquare(m, move));
    if (i === -1) {
      return null;
    }
    rest.splice(i, 1);
  }
  return rest;
}

/** `true` when every pending move belongs to `play` (see `unmatched`). */
export function movesMatchPlay(pending: readonly Move[], play: Play): boolean {
  return unmatched(pending, play) !== null;
}

/** The moves of `play` still to be entered after `pending` (empty when none match). */
export function remainingMoves(pending: readonly Move[], play: Play): Move[] {
  return unmatched(pending, play) ?? [];
}

/**
 * A `Play` argument for `applyPlay` carrying only the moves: the engine
 * accepts a play object without `notation` (`engine/bg-wasm/src/api.rs`),
 * and a partial play has no engine notation to give.
 */
export function partialPlay(moves: readonly Move[]): Play {
  return { moves: [...moves] } as Play;
}
